File: Problems/No.572_Subtree_of_Another_Tree/subtree_of_another_tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):

    def PreOrder(self, r_node, sr_val, psr_nodes):
        if not r_node:
            return
        if (r_node.val == sr_val):
            psr_nodes.append(r_node)
        self.PreOrder(r_node.left, sr_val, psr_nodes)
        self.PreOrder(r_node.right, sr_val, psr_nodes)
    
    def is_same(self, node1, node2):
        if not node1 and not node2:
            return True
        elif not node1 or not node2:
            return False
        elif (node1.val != node2.val):
            return False
        else:
            return self.is_same(node1.left, node2. left) and self.is_same(node1.right, node2.right)

    def isSubtree(self, root, subRoot):
        """
        :type root: TreeNode
        :type subRoot: TreeNode
        :rtype: bool
        """

        # parameter
        potential_subroots = []

        # start computing
        self.PreOrder(root, subRoot.val, potential_subroots)
        for node in potential_subroots:
            if self.is_same(node, subRoot):
                return True
        
        return False

File: Problems/No.572_Subtree_of_Another_Tree/subtree_of_another_tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:

    void PreOrder(TreeNode* r_node, int &sr_val, std::vector<TreeNode *> &psr_nodes) {

        if (r_node == nullptr) return;
        if (r_node->val == sr_val) psr_nodes.push_back(r_node);
        PreOrder(r_node->left, sr_val, psr_nodes);
        PreOrder(r_node->right, sr_val, psr_nodes);

    }

    bool is_same (TreeNode *node1, TreeNode *node2) {

        if ((node1 == nullptr) && (node2 == nullptr)) return true;
        else if ((node1 == nullptr) || (node2 == nullptr)) return false;
        else if (node1->val != node2->val) return false;
        else return (is_same(node1->left, node2->left) && is_same(node1->right, node2->right));

    }

    bool isSubtree(TreeNode* root, TreeNode* subRoot) {
        
        // parameter
        std::vector<TreeNode *> potential_subroots;

        // start computing
        PreOrder(root, subRoot->val, potential_subroots);
        for (auto &node: potential_subroots)
            if (is_same(node, subRoot)) return true;
        
        return false;

    }
};
